packages/SwingSet/src/liveslots/stop-vat.js

```diff
@@ -0,0 +1,393 @@
+import { makeVatSlot, parseVatSlot } from '../lib/parseVatSlots.js';
+
+// This file has tools to run during the last delivery of the old vat
+// version, `dispatch.stopVat()`, just before an upgrade. It is
+// responsible for deleting as much of the non-retained data as
+// possible. The primary function is `releaseOldState()`, at the end;
+// everything else is a helper function.
+
+// The only data that should be retained are those durable objects and
+// collections which are transitively reachable from two sets of
+// roots: the durable objects exported to other vats, and the
+// "baggage" collection. All other durable objects, and *all*
+// merely-virtual objects (regardless of export status) should be
+// deleted. All imports which were kept alive by dropped object should
+// also be dropped.
+
+// However, the possibility of cycles within durable storage means
+// that a full cleanup requires a mark+sweep pass through all durable
+// objects, which I think is too expensive for right now.
+
+// So instead, I'm going to settle on a cheaper `stopVat` which
+// correctly drops durable objects and imports that were only kept
+// alive by 1: RAM, 2: non-durable exports, or 3: non-durable
+// objects/collections. It will require a walk through all non-durable
+// objects and collections, but not a mark+sweep through all durable
+// objects.
+
+// This cheaper form may leak some imports and storage, but should not
+// allow the new vat to access anything it shouldn't, nor allow other
+// vats to cause confusion in the new version (by referencing exports
+// which the vat no longer remembers).
+
+const rootSlot = makeVatSlot('object', true, BigInt(0));
+
+function rejectAllPromises({ m, deciderVPIDs, syscall }) {
+  // Pretend that userspace rejected all non-durable promises. We
+  // basically do the same thing that `thenReject(p, vpid)(rejection)`
+  // would have done, but we skip ahead to the syscall.resolve
+  // part. The real `thenReject` also does pRec.reject(), which would
+  // give control to userspace (who might have re-imported the promise
+  // and attached a .then to it), and stopVat() must not allow
+  // userspace to gain agency.
+
+  const rejectCapData = m.serialize('vat upgraded');
```

Comment:
There's a bit of OCD in me that is bothered by the `'vat upgraded'` text because the vat might not yet have been upgraded when whoever's waiting on the promise receives the rejection.  Kind of like those early Final Fantasy games where you saw text narrating combat in the past tense just before you were shown the animation of the combat.  Maybe `'vat halted for upgrade`' or something?

Reply:
Oh, think of it this way: the only way for anybody to see this message is if the vat was successfully upgraded, because if something failed, all state would get rolled back, including the rejection of those promises. The anthropological argument always wins, as far as you know :).